docs/proposals/service_links.md

```diff
@@ -0,0 +1,231 @@
+# Explicit service links
+
+## Abstract
+
+Proposes an API for explicit links to services in the pod spec.
+
+## Motivation
+
+Currently, all pods are automatically injected with docker-link-style
+environment variables for each service in their namespaces.  Making service
+dependencies explicit in the PodSpec has a number of advantages:
+
+1.  Reduce accidental coupling
+2.  Reduce the number of IP tables rules that have to be created by the
+    kube-proxy
+
+## Constraints and Assumptions
+
+1.  The old behavior must remain the default (for now)
+2.  It must be possible to migrate off the default to using explicit links
```

Comment:
You should also note that we only inject these because we didn't have DNS before.  Most people don't use the ports, but that's somewhat useful.  Anyone can emulate these by setting SERVICE_HOST=SERVICE_NAME and SERVICE_PORT=precanned.

Having a powerful injection tool to inject the actual service IP may not actually be useful, since we now have DNS.